File: 876-hand-of-straights/hand-of-straights.go
/*

1 2 2 3 3 4 6 7 8
*/

import (
    "sort"
)

func isNStraightHand(hand []int, groupSize int) bool {
    if len(hand) % groupSize != 0 {
        return false
    }

    numCounts := make(map[int]int, len(hand))
    for _, num := range hand{
        numCounts[num]++
    }

    sort.Ints(hand)

    for _, num := range hand{
        if numCounts[num] == 0 {continue}

        size := 0
        currNum := num
        for size < groupSize{
            if numCounts[currNum] == 0{
                return false
            }
            numCounts[currNum]--
            currNum++
            size++
        }
    }
    return true
    
}

File: 876-hand-of-straights/hand-of-straights.py
"""
1 2 2 3 3 4 6 7 8
    s     
    i

size: 3

1 2 3
"""

class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        if len(hand) % groupSize != 0:
            return False
        
        counts = defaultdict(int)
        for num in hand:
            counts[num] += 1
        
        hand.sort()
        
        for num in hand:
            if counts[num] == 0:
                continue
            
            curr = num
            curr_size = 1
            counts[curr] -= 1

            while counts[curr + 1] > 0 and curr_size < groupSize:
                curr_size += 1
                counts[curr + 1] -= 1
                curr += 1
            
            if curr_size < groupSize:
                return False
            

        return True
                